CMakeLists.txt

```diff
@@ -0,0 +1,161 @@
+# etc.
+#
+find_package(Torch REQUIRED)
+
+#
+# Set up GPU language and check the torch version and warn if it isn't
+# what is expected.
+#
+if (NOT HIP_FOUND AND CUDA_FOUND)
+    set(VLLM_GPU_LANG "CUDA")
+
+    # Check CUDA is at least 11.6
+    if (CUDA_VERSION VERSION_LESS 11.6)
+        message(FATAL_ERROR "CUDA version 11.6 or greater is required.")
+    endif ()
+
+    if (NOT Torch_VERSION VERSION_EQUAL ${TORCH_SUPPORTED_VERSION_CUDA})
+        message(WARNING "Pytorch version ${TORCH_SUPPORTED_VERSION_CUDA} "
+                "expected for CUDA build, saw ${Torch_VERSION} instead.")
+    endif ()
+elseif (HIP_FOUND)
+    set(VLLM_GPU_LANG "HIP")
+
+    # Importing torch recognizes and sets up some HIP/ROCm configuration but does
+    # not let cmake recognize .hip files. In order to get cmake to understand the
+    # .hip extension automatically, HIP must be enabled explicitly.
+    enable_language(HIP)
+
+    # ROCm 5.X and 6.X
+    if (ROCM_VERSION_DEV_MAJOR GREATER_EQUAL 5 AND
+            NOT Torch_VERSION VERSION_EQUAL ${TORCH_SUPPORTED_VERSION_ROCM})
+        message(WARNING "Pytorch version >= ${TORCH_SUPPORTED_VERSION_ROCM} "
+                "expected for ROCm build, saw ${Torch_VERSION} instead.")
+    endif ()
+else ()
+    message(FATAL_ERROR "Can't find CUDA or HIP installation.")
+endif ()
+
+#
+# Override the GPU architectures detected by cmake/torch and filter them by
+# the supported versions for the current language.
+# The final set of arches is stored in `VLLM_GPU_ARCHES`.
+#
+override_gpu_arches(VLLM_GPU_ARCHES
+        ${VLLM_GPU_LANG}
+        "${${VLLM_GPU_LANG}_SUPPORTED_ARCHS}")
+
+#
+# Query torch for additional GPU compilation flags for the given
+# `VLLM_GPU_LANG`.
+# The final set of arches is stored in `VLLM_GPU_FLAGS`.
+#
+get_torch_gpu_compiler_flags(VLLM_GPU_FLAGS ${VLLM_GPU_LANG})
+
+#
+# Set nvcc parallelism.
+#
+if (NVCC_THREADS AND VLLM_GPU_LANG STREQUAL "CUDA")
+    list(APPEND VLLM_GPU_FLAGS "--threads=${NVCC_THREADS}")
+endif ()
+
+
+#
+# _C extension
+#
+
+# TODO(luka): happy to hardcode, but these are generated so globbing seems fine.
+file(GLOB FLASH_ATTN_GEN_SRCS "csrc/flash_attn/src/flash_fwd_*.cu")
+message(DEBUG "FLASH_ATTN_GEN_SRCS: ${FLASH_ATTN_GEN_SRCS}")
+
+define_gpu_extension_target(
+        _C
+        DESTINATION vllm_flash_attn
+        LANGUAGE ${VLLM_GPU_LANG}
+        SOURCES csrc/flash_attn/flash_api.cpp ${FLASH_ATTN_GEN_SRCS}
+        COMPILE_FLAGS ${VLLM_GPU_FLAGS}
+        ARCHITECTURES ${VLLM_GPU_ARCHES}
+        # pybind does not support SOABI
+        # USE_SABI 3
+        WITH_SOABI
+)
+
+# TODO(luka): remove this once PyBind is removed
+find_library(TORCH_PYTHON_LIBRARY torch_python PATHS "${TORCH_INSTALL_PREFIX}/lib" REQUIRED)
+target_link_libraries(_C PRIVATE ${TORCH_PYTHON_LIBRARY})
+
+target_include_directories(_C PRIVATE
+        csrc/flash_attn
+        csrc/flash_attn/src
+        csrc/cutlass/include)
```

Comment:
You could use `INCLUDE_DIRECTORIES` on `define_gpu_extension_target` to set these.

Reply:
To me this is clearer, and it doesn't require redundant proxy variables, and doesn't duplicate functionality (and is `INCLUDE_DIRECTORIES` on the GPU target `PRIVATE` or `PUBLIC` by default)? I'm happy to change it though if you have a strong preference. I switched the other flags over to avoid repeating the generator expression

Reply:
It defaults to `PRIVATE`.  I don't have super strong feelings about either style.

Reply:
Yeah it was more of a rhetorical question 😅 . I'll keep this one